File: sample/src/main/java/com/readdle/swiftjava/sample/FloatTest.kt
package com.readdle.swiftjava.sample

import com.readdle.codegen.anotation.*
import java.lang.annotation.Native

@SwiftValue
data class FloatTestStruct(var zero: Float = 0f,
                           var infinity: Float = Float.POSITIVE_INFINITY,
                           var negativeInfinity: Float = Float.NEGATIVE_INFINITY,
                           var optional: Float? = 0f,
                           var optionalNil: Float? = null)

@SwiftReference
class FloatTest private constructor() {

    @SwiftDelegate(protocols = ["FloatTestParamProtocol"])
    interface FloatParamProtocol {
        @SwiftCallbackFunc
        fun testParam(param: Float): Boolean
    }

    @SwiftDelegate(protocols = ["FloatTestReturnTypeProtocol"])
    interface FloatReturnTypeProtocol {
        @SwiftCallbackFunc
        fun testReturnType(): Float
    }

    @SwiftDelegate(protocols = ["FloatTestOptionalParamProtocol"])
    interface FloatOptionalParamProtocol {
        @SwiftCallbackFunc
        fun testOptionalParam(param: Float?): Boolean
    }

    @SwiftDelegate(protocols = ["FloatTestOptionalReturnTypeProtocol"])
    interface FloatOptionalReturnTypeProtocol {
        @SwiftCallbackFunc
        fun testOptionalReturnType(): Float?
    }

    @SwiftBlock("(Float) -> Float")
    interface FloatBlock {
        fun call(value: Float): Float
    }

    @SwiftBlock("(Float?) -> Float?")
    interface OptionalFloatBlock {
        fun call(value: Float?): Float?
    }

    companion object {
        @JvmStatic
        external fun testZero(): Float

        @JvmStatic
        external fun testInfinite(): Float

        @JvmStatic
        external fun testNan(): Float

        @JvmStatic
        external fun testParam(param: Float): Boolean

        @JvmStatic
        external fun testReturnType(): Float

        @JvmStatic
        external fun testOptionalParam(param: Float?): Boolean

        @JvmStatic
        external fun testOptionalReturnType(): Float?

        @JvmStatic
        external fun testProtocolParam(callback: FloatParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolReturnType(callback: FloatReturnTypeProtocol): Float

        @JvmStatic
        external fun testProtocolOptionalParam(callback: FloatOptionalParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolOptionalReturnType(callback: FloatOptionalReturnTypeProtocol): Float?

        @JvmStatic
        external fun testEncode():  FloatTestStruct

        @JvmStatic
        external fun testDecode(value: FloatTestStruct): Boolean

        @JvmStatic
        external fun testBlock(@SwiftBlock block: FloatBlock): Boolean

        @JvmStatic
        external fun testOptionalBlock(@SwiftBlock block: OptionalFloatBlock): Boolean
    }

    @Native
    var nativePointer: Long = 0

    external fun release()

}

File: sample/src/main/swift/Sources/SampleAppCore/FloatTest.swift
//
// Created by Andrew on 2/4/18.
//

import Foundation

public struct FloatTestStruct: Codable, Hashable {
    public var zero: Float = Float.zero
    public var infinity: Float = Float.infinity
    public var negativeInfinity: Float = -1 * Float.infinity
    public var optional: Float? = Float.zero
    public var optionalNil: Float? = nil
}

public protocol FloatTestParamProtocol {
    func testParam(_ param: Float) -> Bool
}

public protocol FloatTestReturnTypeProtocol {
    func testReturnType() -> Float
}

public protocol FloatTestOptionalParamProtocol {
    func testOptionalParam(_ param: Float?) -> Bool
}

public protocol FloatTestOptionalReturnTypeProtocol {
    func testOptionalReturnType() -> Float?
}

public typealias FloatBlock = (_ value: Float) -> Float
public typealias OptionalFloatBlock = (_ value: Float?) -> Float?

public class FloatTest {

    public static func testZero() -> Float {
        return 0
    }

    public static func testInfinite() -> Float {
        return Float.infinity
    }

    public static func testNan() -> Float {
        return Float.nan
    }

    public static func testParam(_ param: Float) -> Bool {
        return param == Float.infinity
    }

    public static func testReturnType() -> Float {
        return Float.infinity
    }

    public static func testOptionalParam(_ param: Float?) -> Bool {
        return param == Float.infinity
    }

    public static func testOptionalReturnType() -> Float? {
        return Float.infinity
    }

    public static func testProtocolParam(_ callback: FloatTestParamProtocol) -> Bool {
        return callback.testParam(Float.infinity)
    }

    public static func testProtocolReturnType(_ callback: FloatTestReturnTypeProtocol) -> Float {
        return callback.testReturnType()
    }

    public static func testProtocolOptionalParam(_ callback: FloatTestOptionalParamProtocol) -> Bool {
        return callback.testOptionalParam(Float.infinity)
    }

    public static func testProtocolOptionalReturnType(_ callback: FloatTestOptionalReturnTypeProtocol) -> Float? {
        return callback.testOptionalReturnType()
    }

    public static func testEncode() ->  FloatTestStruct {
        return FloatTestStruct()
    }

    public static func testDecode(_ value: FloatTestStruct) -> Bool {
        return value == FloatTestStruct()
    }

    public static func testBlock(_ block: FloatBlock) -> Bool {
        let value = block(0)
        return value == 0
    }

    public static func testOptionalBlock(_ block: OptionalFloatBlock) -> Bool {
        let value = block(nil)
        return value == nil
    }

}